debug_module.tex

```diff
@@ -222,6 +210,21 @@ \section{Abstract Commands} \label{abstractcommands}
     \end{tabulary}
 \end{table}
 
+\begin{commentary}
+    The Abstract Command interface is designed to allow a debugger to write
+    commands as fast as possible, and then later check whether they completed
+    without error.  In the common case the debugger will be much slower than
+    the target and commands succeed, which allows for maximum throughput. If
+    there is a failure, the interface ensures that no commands execute after
+    the failing one.  To discover which command failed, the debugger has to
+    look at the state of the DM (eg. contents of \Rcommand) or hart (eg.
+    contents of a register modified by a Program Buffer program) to determine
+    which one failed.
+\end{commentary}
+
+If a hart is reset while executing an abstract command, then the abstract
```

Comment:
"If a hart is reset..." do you mean via a write to \Fhartsel, or by any means including a non-debugger controlled system wide reset? 

This is a backwards incompatible change to add this requirement. Also, the error doesn't really make sense. We don't really have abstract commands fail when a hart is in reset, so far. Maybe THAT's the bug, but maybe you want to be able to read registers in a hart while it's in reset...

Reply:
The way I imagine this working is that the DM knows that an abstract command is executing, and if it sees havereset or something go high during that time, it knows a reset occurred during that time. In that case the DM can assume that the abstract command execution got lost,  so it sets error, and the debugger can try again, etc.

This all assumes that the hart is the thing that ends up executing the abstract command. If the DM is doing it all, then it still seems fairly straightforward to implement.

The only alternative I can think of is that somehow the abstract command completes even though the hart got reset. That seems potentially really complex to implement.

Reply:
I don't think we should *mandate* that the command has to end in error if the hart is reset. If an implementation supports a sideband read of a register, then that may and should work if the hart is in reset, not just give an error. This is useful functionality in general.

I think it's a separate case as to the debug module realizing that an abstract command will never complete because a hart is reset, but that is not ALWAYS the case.

Reply:
Perhaps a better solution is that the debugger MAY report the "wrong state" error, or some new error meaning "we're not sure if that worked or not". The previously-named "halt/resume" error was a clear indicator that the debug module didn't do the command at all, because the hart wasn't in the right state. This is more of "something went wrong while we were doing that so we don't know what happened" sort of case.

Reply:
The main thing I want to prevent is abstract commands/the DM hanging due to a hart being reset at the wrong time. Currently we can abort a hanging abstract command by resetting the hart. If you allow abstract commands while reset is asserted, then I don't see how you can kick a hart that's hung in an abstract command. (Abstract commands might hang because of an infinite loop in the Program Buffer, or perhaps if a reset causes some protocol problems between the hart and DM.)

I'm not against allowing register access during reset, but I don't think it's essential, since a debugger can always access registers immediately out of reset and then access them.

I'm fine adding an error code that an implementation may set if an abstract command was interrupted by reset, but also giving implementations the option to have the command complete (possibly returning bogus results).

Reply:
I agree with pretty much all of this, that we need some elegant way to cancel a command or have it cancel itself if something bad is detected. But I do think we should keep in mind the "abstract" part of the abstract command--- we shouldn't assume it has anything to do with a hart's execution since in theory halt is supposed to be able to do anything to a hart, like completely gate its clock, to allow registers to be examined.

>"If you allow abstract commands while reset is asserted, then I don't see how you can kick a hart that's hung in an abstract command. "

I think you have it a little backwards... what you are really asking for is a way to cancel the abstract command. If you allow hart reset while a abstractcs.busy=1, then you can reset the hart. But what we don't have right now is a way to hard-clear abstrtactcs.busy. Saying that hartreset is the way to do that seems a little weird. Why not just have a "kill abstractcs, and I don't care what the consequences are"? Either way, it's going to potentially screw up the debug module's view of the world, so you may have to resort to dmactive level reset.

Reply:
Has there been consensus on unexpected reset? 
The other case is a hanging abstract command (without unexpected havereset).
That could be the result of a HW issue, or it could be SW (e.g. the abstract command is running program buffer code spinning on a lock that has been lost). The former case is fatal; (it should never happen) and be the resultof a system wide hang, in which case even a directed hartreset isn't good enough. 
The latter is livelock, not deadlock - but there is no way to determine whether that is the case or not. Even if you could cancel the command (and I would suggest writing a command of allzero be the way to signal that - register_access w/no transfer and no postexec) you can't know if it were livelock or deadlock and must assume the worst. 

And, you can't "cancel" an external transaction;  it leaves microachitectural state hanging around that you have no control over which will cause hangs further down the road.
So: my conclusion is that a hanging abstract command is fatal, and a directed hartreset is the least you can do to recover.